src/databricks/sql/utils.py

```diff
@@ -387,26 +393,121 @@ def inject_parameters(operation: str, parameters: Dict[str, str]):
+    """Check for the presence of `%s` in the operation string.
+    If it's present, the operation may contain positional query markers. Return True.
+
+
+    If `operation` contains no instances of `%s` then return False.
+
+    ```
+    "SELECT * FROM table WHERE field = %s and other_field = %s"
+    ```
+
+    Yields
+
+    ```
+    SELECT * FROM table WHERE field = ? and other_field = ?
+    ```
+
+    Note that this function doesn't parse the contents of operation! So if you pass a query like
+
+    ```sql
+    SELECT * FROM table WHERE field LIKE '%some_string'
+    ```
+
+    The output will be invalid SQL:
+
+    ```sql
+    SELECT * FROM table WHERE field LIKE '?ome_string'
+    ```
+
+    This is a limitation of conflicting Python Format String Mini Language and SQL syntax
+    has always affected the INLINE parameter approach to string interpolation. To avoid it,
+    the operation should be written to use `?` directly and not `%s`.
+    """
+
+    interpolated = operation.replace("%s", "?")
+    return interpolated == operation
+
+
+def _interpolate_named_markers(
+    operation: str, parameters: Union[ListOfParameters, DictOfParameters]
+) -> str:
+    """Replace all instances of `%(param)s` in `operation` with `:param`.
+
+    If `operation` contains no instances of `%(param)s` then the input string is returned unchanged.
+
+    ```
+    "SELECT * FROM table WHERE field = %(field)s and other_field = %(other_field)s"
+    ```
+
+    Yields
+
+    ```
+    SELECT * FROM table WHERE field = :field and other_field = :other_field
+    ```
+    """
+
+    try:
+        return operation % _generate_named_interpolation_values(parameters)
+    except TypeError:
+        # TypeError occurs if there are no %(param)s markers in the operation
+        return operation
+
+
+def transform_paramstyle(
+    operation: str,
+    parameters: Union[ListOfParameters, DictOfParameters],
+    param_structure: ParameterStructure,
+) -> str:
     """
     Performs a Python string interpolation such that any occurence of `%(param)s` will be replaced with `:param`
+    If param_structure == ParameterStructure.POSITIONAL, then the markers like `%s` will be replaced with `?` instead.
 
     This utility function is built to assist users in the transition between the default paramstyle in
     this connector prior to version 3.0.0 (`pyformat`) and the new default paramstyle (`named`).
 
     This method will fail if parameters is passed as a list.
 
     Args:
-        operation (str): The operation or SQL text to transform.
-        parameters (Dict[str, Any]): The parameters to use for the transformation.
+        operation: The operation or SQL text to transform.
+        parameters: The parameters to use for the transformation.
 
     Returns:
         str
     """
+    output = operation
+    if (
+        param_structure == ParameterStructure.POSITIONAL
+        and _may_contain_inline_positional_markers(operation)
+    ):
+        logger.warning(
```

Comment:
I'm not sure we really need this log message.

Reply:
If we're going to try to support %s at all, then yes i think we need this message.  As mentioned above, I would instead suggest we explicitly not support it due to conflicts with our like syntax.